### ISSUE TITLE
Implement randomized compiler based on NuOps and bring GateTabulation to Cirq core

### ISSUE THREAD

**Issue by tanujkhattar** [quantumlib/Cirq#4362]
**Is your feature request related to a use case or problem? Please describe.**
Randomized compilation to arbitrary target gatesets is a useful compilation primitive for the NISQ era. Currently, cirq only has GateTabulation method in cirq_google which supports approximate compilation to arbitrary target gatesets. However, being in cirq_google, GateTabulation is not very easily discoverable. 

There are also other nice randomized compilation techniques like https://arxiv.org/pdf/2106.15490.pdf which should exist in cirq_core. 

**Describe the solution you'd like**
- The GateTabulation compilation technique is not specific to cirq_google and should exist in cirq_core. There have also been requests in the past to implement a generic optimizer based on GateTabulation (#4059), which should also exist in cirq_core.
- The NuOps compilation method described https://arxiv.org/pdf/2106.15490.pdf is very similar to GateTabulation, but smarter in the sense that instead of first trying to tabulate the weyl chamber by generating random single qubit unitaries it rather considers the rotation angles of single qubit unitaries as optimization parameters and uses a numerical optimizer to find those angles for a given target unitary. This removes the heavy precomputation stage needed for GateTabulation and should potentially also improve the accuracy. It also allows for noise adoptability during randomized gate compilation by enabling compilation to different base gates for different target unitaries (which is still possible by GateTabulation but would need a precomputation stage for each different base gate). 

**What is the urgency from your perspective for this issue? Is it blocking important work?**
This work should be done as part of the gate compiling roadmap item (#3239) and we should eventually aim for a suit of randomized compilation methods similar to the various exact kak based analytical gate synthesis methods that exist today.

P2 - we should do it in the next couple of quarters

**Comment by Ashalynd** [comment]
Hi Tanuj. Would you mind if I have a look at this?

**Comment by tanujkhattar** [comment]
Sure! There are multiple smaller work items for this larger bug
1. Bring GateTabulation out of cirq-google to cirq-core
2.Add a generic optimizer that makes use of GateTabulation
3. Implement the NuOps compiler, given in the linked paper, in cirq-core.

[@chrischris96](https://github.com/chrischris96)  had also reached out to me over email earlier expressing interest in this issue but I think the issue is large enough that more than one person can work on it in parallel. 

Please let me know if you have any further questions.

**Comment by Ashalynd** [comment]
Thanks for the quick reply!

Is there a dependency between the first (GateTabulation refactoring + optimizer) and the second (NuOps implementation) tasks?

If not, would it help to split this issue into two or more separate ones?

**Comment by tanujkhattar** [comment]
No, there's no dependency. 

There's already https://github.com/quantumlib/Cirq/issues/4059 that talks about adding an optimizer that uses the GateTabulation method. Feel free to create new smaller sub-issues to track specific work items.

**Comment by chrischris96** [comment]
@Ashalynd I read the paper and don't have any preferences about which part to work on. Would you rather work on the optimizer or the NuOps implementation task?

**Comment by Ashalynd** [comment]
Hi @chrischris96,

Sure, I can take the first part (refactoring GateTabulation and creating an optimizer). 

@tanujkhattar  - that would correspond to issue #4059 plus may be a separate issue for refactoring, correct?

**Comment by tanujkhattar** [comment]
@Ashalynd Yes, that's correct.

**Comment by Ashalynd** [comment]
I created #4461 (can't assign additional  labels to it though).

**Comment by tanujkhattar** [comment]
@chrischris96 As part of tracking the progress towards our Cirq 1.0 roadmap items, I wanted to check in on the progress on this issue. Could you start working on it yet?

**Comment by chrischris96** [comment]
@tanujkhattar I am having exams until Monday but will start working on it full-time next week

**Comment by Ashalynd** [comment]
What is the status of the next task (write generic optimizer based on GateTabulation)? @tanujkhattar 
I have more cycles now , can work on that if it's still in scope.

**Comment by Ashalynd** [comment]
Gentle ping :) Is this issue still in scope? If yes, is #4059 taken / are there any parts left for this issue?

**Comment by tanujkhattar** [comment]
@Ashalynd You can work on #4059 now that GateTabulation has been moved to cirq-core and we have new transformer API and transformer primitives checked-in, which should be used to implement #4059

Please let me know if you have any questions.

**Comment by Ashalynd** [comment]
Thanks for confirmatilon!

**Comment by ToastCheng** [comment]
Hi! If there is no active progress, I can take this one to continue the NuOp implementation.

**Comment by mhucka** [comment]
@ToastCheng Thank you for offering! It's assigned to you.

**Comment by ToastCheng** [comment]
I followed the tabulation's pattern and added NuOp support to Cirq (#8273). To verify the implementation, I also added unit tests to check if it reproduces the paper's testing result.

I'm not sure whether we should also add a gateset for NuOp so that users can use it with optimize_for_target_gateset. Although many gatesets seem to be device-specific (e.g. SycamoreTargetGateset), there are also general ones like CZTargetGateset. If that would be helpful, I can draft a separate PR for that.

**Comment by ToastCheng** [comment]
Given that the change could be complex, I drafted a [mini RFC](https://docs.google.com/document/d/10xPjUO_Yeflv6oMo91HS2xusSf3lhOZfvqXp81F9NTQ/edit?usp=sharing) to make it clearer.

**Comment by ToastCheng** [comment]
Hi! Some updates:

I tried to add a gateset to support using NuOp via `optimize_for_target_gateset` (currently a [commit](https://github.com/ToastCheng/Cirq/commit/5c53d8ce2df0c6a8e24476de2e0af504ca8182d6) in my tree), and ran benchmarks comparing NuOp with Cirq's tabulation. Details are in the [Experiment and benchmark](https://docs.google.com/document/d/10xPjUO_Yeflv6oMo91HS2xusSf3lhOZfvqXp81F9NTQ/edit?tab=t.0#bookmark=id.dvmfx9yjwslv) section. If this direction looks right, I can add the change either to #8273 or a separate PR.
